### PR TITLE
feat(autocomplete): add `completion` script to install autocomplete

### PR DESCRIPTION
**Pull request by kentcdodds**
**What**: This adds a command called `completion` which installs the
autocomplete shell script. It also adds the logic to provide the
autocompletion.

**Why**: Because that's pretty swell!

**How**: Using omelette which is pretty awesome.

@tleunen, @SamVerschueren, and @jamischarles, do ya'll mind trying this out?

```
npm install --global p-s@beta
nps completion <path-to-your-bash-profile-file>
nps <tab>
```

`<path-to-your-bash-profile-file>` defaults to `~/.bash_profile`.

I'd appreciate a code review as well :)

### CODE REVIEW

**Comment by kentcdodds** [comment]
LGTM

**Comment by kentcdodds** [comment]
Whoops, forgot to reference the original issue in my PR. It's #28

**Comment by kentcdodds** [comment]
For some reason it's not working for me. [This line](https://github.com/kentcdodds/p-s/blob/pr/autocomplete/src/bin-utils/autocomplete/index.js#L19) just stops the program without any error :-( Still investigating.

**Comment by kentcdodds** [comment]
Maybe we should add something to detect whether this `p-s` is installed globally (use [this](https://www.npmjs.com/package/detect-installed)?), because installing completion without a global install would be kinda worthless...

**Comment by kentcdodds** [comment]
Sweet! I got it working! Turns out that omelette only supports `$SHELL`s containing `bash` and `zsh`. Mine was set to `/bin/sh` somehow. After [fixing that](https://twitter.com/kentcdodds/status/759133923989295104) the install worked flawlessly!

Would like to hear feedback from someone using `zsh` first.

**Comment by edm00se** [comment]
It looks like it didn't like my using it in `zsh`, from `.zshrc`.

p-s_completion_zsh

Error message on double-clicking "tab". I am using `nvm` and a v4.x LTS (my default).

> λ Alfheim ~ → nps /Users/thor/.nvm/versions/node/v4.4.5/lib/node_modules/p-s/dist/bin-utils/index.js:123
>   return moduleName[0] === '.' ? (0, _path.resolve)(process.cwd(), moduleName) : moduleName;
>                    ^
> 
> TypeError: Cannot read property '0' of null
>     at getModuleRequirePath (/Users/thor/.nvm/versions/node/v4.4.5/lib/node_modules/p-s/dist/bin-utils/index.js:123:20)
>     at attemptModuleRequire (/Users/thor/.nvm/versions/node/v4.4.5/lib/node_modules/p-s/dist/bin-utils/index.js:128:23)
>     at getPSConfig (/Users/thor/.nvm/versions/node/v4.4.5/lib/node_modules/p-s/dist/bin/p-s.js:82:41)
>     at Object.<anonymous> (/Users/thor/.nvm/versions/node/v4.4.5/lib/node_modules/p-s/dist/bin/p-s.js:49:18)
>     at Module._compile (module.js:409:26)
>     at Object.Module._extensions..js (module.js:416:10)
>     at Module.load (module.js:343:32)
>     at Function.Module._load (module.js:300:12)
>     at Function.Module.runMain (module.js:441:10)
>     at startup (node.js:139:18)
> /Users/thor/.nvm/versions/node/v4.4.5/lib/node_modules/p-s/dist/bin-utils/index.js:123
>   return moduleName[0] === '.' ? (0, _path.resolve)(process.cwd(), moduleName) : moduleName;
>                    ^
> 
> TypeError: Cannot read property '0' of null
>     at getModuleRequirePath (/Users/thor/.nvm/versions/node/v4.4.5/lib/node_modules/p-s/dist/bin-utils/index.js:123:20)
>     at attemptModuleRequire (/Users/thor/.nvm/versions/node/v4.4.5/lib/node_modules/p-s/dist/bin-utils/index.js:128:23)
>     at getPSConfig (/Users/thor/.nvm/versions/node/v4.4.5/lib/node_modules/p-s/dist/bin/p-s.js:82:41)
>     at Object.<anonymous> (/Users/thor/.nvm/versions/node/v4.4.5/lib/node_modules/p-s/dist/bin/p-s.js:49:18)
>     at Module._compile (module.js:409:26)
>     at Object.Module._extensions..js (module.js:416:10)
>     at Module.load (module.js:343:32)
>     at Function.Module._load (module.js:300:12)
>     at Function.Module.runMain (module.js:441:10)
>     at startup (node.js:139:18)
> /Users/thor/.nvm/versions/node/v4.4.5/lib/node_modules/p-s/dist/bin-utils/index.js:123
>   return moduleName[0] === '.' ? (0, _path.resolve)(process.cwd(), moduleName) : moduleName;
>                    ^
> 
> TypeError: Cannot read property '0' of null
>     at getModuleRequirePath (/Users/thor/.nvm/versions/node/v4.4.5/lib/node_modules/p-s/dist/bin-utils/index.js:123:20)
>     at attemptModuleRequire (/Users/thor/.nvm/versions/node/v4.4.5/lib/node_modules/p-s/dist/bin-utils/index.js:128:23)
>     at getPSConfig (/Users/thor/.nvm/versions/node/v4.4.5/lib/node_modules/p-s/dist/bin/p-s.js:82:41)
>     at Object.<anonymous> (/Users/thor/.nvm/versions/node/v4.4.5/lib/node_modules/p-s/dist/bin/p-s.js:49:18)
>     at Module._compile (module.js:409:26)
>     at Object.Module._extensions..js (module.js:416:10)
>     at Module.load (module.js:343:32)
>     at Function.Module._load (module.js:300:12)
>     at Function.Module.runMain (module.js:441:10)
>     at startup (node.js:139:18)
> 
>

**Comment by edm00se** [comment]
It's not liking me (or something I'm doing). Running node v4.x LTS (my default) from nvm.

screen shot 2016-07-29 at 9 03 09 pm

**Comment by kentcdodds** [comment]
Hmmm... do you think you could do some digging for me @edm00se? First, make sure that the completion was actually installed in your config. Then go into the globally installed `p-s` and find this file and add some console logs around [here](https://github.com/kentcdodds/p-s/blob/7b8b04ce77daca01e35ac6ca61593c8eb31167ef/src/bin/p-s.js#L36) and [here](https://github.com/kentcdodds/p-s/blob/7b8b04ce77daca01e35ac6ca61593c8eb31167ef/src/bin/p-s.js#L38).

It should be taking the first codepath...

**Comment by edm00se** [comment]
I cloned the repo, checked out `pr/autocomplete`, made these `console.log` additions, and ran `npm start build` and `npm link`.

screen shot 2016-07-29 at 9 20 56 pm

screen shot 2016-07-29 at 9 30 10 pm

I cleared the last few lines of my `.zshrc` then re-ran `nps completion`. Completion did append to my `.zshrc`.
screen shot 2016-07-29 at 9 34 15 pm

Here's what it looked like adding the completion after all that (double-clicked tab after entering `nps s`, for brevity):

> λ Alfheim transformers-names → λ git master → nps sScripts must resolve to strings. There is no script that can be resolved from "nps" https://github.com/kentcdodds/p-s/blob/v0.0.0-semantically-released/other/ERRORS_AND_WARNINGS.md#missing-script
>                                               nps s
> script\",\"script\":\"codecov\"\},\"release\":\{\"description\":\"We
> scripts
> semantic-release
> semantic-release.
> several
> shoudlRun,
> should
> so
> src\"\},\"umd\":\{\"description\":\"build
> stats
> stuff\",\"script\":\"p-s
> sure

**Comment by kentcdodds** [comment]
Huh, well I'm totally stumped :-(

**Comment by edm00se** [comment]
This is one of those moments I debate whether I've turned my dotfiles and `.zshrc` into swiss cheese. I wonder how it would perform under a more 'vanilla' zsh environment.

**Comment by kentcdodds** [comment]
Ping @tleunen, @SamVerschueren, and @jamischarles. Could you give this a look and let me know if it works for you?

**Comment by edm00se** [comment]
I got the same in a pretty clean Ubuntu 16.04 vm, w/ zsh, oh-my-zsh, and Node 4.x. :-/ This is entirely strange; omelette.js [shows it working in zsh](https://github.com/f/omelette#omelettejs-v031) and claims to only need node >= 0.8.

screenshot 2016-08-01 11 58 23

File: src/bin-utils/autocomplete/index.test.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  t.true(setupShellInitFileSpy.calledOnce)
+  t.is(actualDestination, destination)
+})
+
+function getAutocomplete() {
+  const onSpy = spy()
+  const initSpy = spy()
+  const omelette = spy(() => {
+    return {on: onSpy, init: initSpy}
+  })
```
Comment (SamVerschueren) [inline src/bin-utils/autocomplete/index.test.js]: You can write this a little bit shorter (if you want off course :))

```js
const omelette = spy(() => ({on: onSpy, init: initSpy}))
```

**Comment by SamVerschueren** [comment]
Works for me 👍

File: src/bin/p-s.js [hunk 2]
```diff
@@ -73,9 +83,30 @@ function onInit() {
...
 }
 
 function onHelp() {
   shouldRun = false
   log.info(help(getPSConfig()))
 }
+
+function onRequestToInstallCompletion() {
+  shouldRun = false
+  const [,,, destination] = process.argv
```
Comment (SamVerschueren) [inline src/bin/p-s.js]: Might be cleaner to just write `const destination = process.argv[3]`

File: src/bin/p-s.js [hunk 3]
```diff
@@ -73,9 +83,30 @@ function onInit() {
...
+  if (destination) {
+    log.info(`Installing p-s autocomplete into ${destination}`)
+  } else {
+    log.info('Installing p-s autocomplete into the default for your current terminal')
+  }
+  log.info(
+    `You're going to need to either resource that file, or open a new instance of ` +
```
Comment (kentcdodds) [inline src/bin/p-s.js]: `resource` should be `re-source`

File: src/bin/p-s.js [hunk 4]
```diff
@@ -73,9 +83,30 @@ function onInit() {
     'Check out your scripts in there. Go ahead and update them and add descriptions to the ones that need it'
   ))
   log.info(colors.gray('Your package.json scripts have also been updated. Run `npm start -- --help` for help'))
+  log.info(colors.gray(
```
Comment (SamVerschueren) [inline src/bin/p-s.js]: It's possible to use template literals for this multiline string and use [redent](https://github.com/sindresorhus/redent). Might be cleaner but it's personal taste. Adds an extra dependency though.

**Comment by kentcdodds** [comment]
@SamVerschueren are you using Bash or zsh?

**Comment by SamVerschueren** [comment]
@kentcdodds Bash

File: src/bin/p-s.js (hunk 4 above)
Comment (kentcdodds) [inline src/bin/p-s.js]: I thought about that, but decided not to... Maybe I should. Would be slightly easier to maintain.

**Comment by kentcdodds** [comment]
I'm going to go ahead and merge this. We'll deal with issues as they come up.

**Pull request merged**